### ISSUE TITLE
Is there a plan to support a range of ports in one instance of go-mmproxy?

### ISSUE THREAD

**Issue by catalystsitaly** [path-network/go-mmproxy#5]
There are tens of backend program running on my server. So if I plan to use go-mmproxy for them, I need to manully start tens of go-mmproxy for each of the backend program.

Is there a way to support a range of ports in one instance of go-mmproxy?

Thanks!

**Comment by kzemek** [comment]
In this scenario, would you want go-mmproxy to run (with a single external proxy port) as a round-robin load-balancer over the backend ports, or do you want to specify 1:1 mappings from proxy ports to backend ports?

**Comment by catalystsitaly** [comment]
> In this scenario, would you want go-mmproxy to run (with a single external proxy port) as a round-robin load-balancer over the backend ports, or do you want to specify 1:1 mappings from proxy ports to backend ports?

I would like to use 1:1 port mapping from proxy ports to backend ports. That is one go-mmproxy instance could handle multiple backend ports and listening to multiple ports accordingly .

**Comment by eneumann-comp** [comment]
> 
> 
> > In this scenario, would you want go-mmproxy to run (with a single external proxy port) as a round-robin load-balancer over the backend ports, or do you want to specify 1:1 mappings from proxy ports to backend ports?
> 
> I would like to use 1:1 port mapping from proxy ports to backend ports. That is one go-mmproxy instance could handle multiple backend ports and listening to multiple ports accordingly .

I'm also interested in this option as well. I am trying to support multiple FTPS servers behind an LB.

**Comment by prostudiocorp** [comment]
Yes, i also need this function, i have thousands of port need be lisntened. i can sponsor for this function,

**Comment by MikoyChinese** [comment]
Or if I have mulit ports to listen, such as 20001, 20003, 13889, how can I do?

**Comment by shshekhar93** [comment]
I needed something similar (to be used in conjunction with rathole, which already supports multiple services in a single process). Starting multiple processes, one for each service, is a reasonable work around, but seemed like a maintenance headache. I've attempted a first cut at implementing support for multiple services in single process. If anyone is still looking for this feature, you can take a look at [my fork](https://github.com/shshekhar93/go-mmproxy).

@kzemek , are you accepting PRs to this repo, I'd much rather merge the changes here than maintain a copy. Unfortunately there ended up being [quite a bit of changes](https://github.com/path-network/go-mmproxy/compare/master...shshekhar93:go-mmproxy:master), most of it is just to pass the server options around instead of reading it from global var. If required, I'd be happy to make any other changes that help the case for merging.

**Comment by volas** [comment]
Currently you can also use systemd templates for better support and automation. For me it looks like
```
systemctl enable mmproxy@port1.service
systemctl enable mmproxy@port2.service
systemctl enable mmproxy@port3.service
```

with ``mmproxy@.service`` template file similar to https://github.com/path-network/go-mmproxy/blob/master/go-mmproxy.service.example, just need to map ports based on the argument.

**Comment by kzemek** [comment]
@shshekhar93 I am accepting PRs at https://github.com/kzemek/go-mmproxy where I'm continuing to maintain the project.

I do think that the "listening on thousands of ports" is a compelling usecase, but handling it properly would require rearchitecturing the project to e.g. run on epoll instead of a goroutine-per-connection model it is now - at which point I'd rather rewrite mmproxy in a more epoll-friendly language than Go.

I don't think there's much value in having this feature as a bolt on to the current architecture, or for use cases of 1-to-1 mapping of tens of ports and not thousands. In these cases I'd recommend starting tens of go-mmproxy processes instead, with something like supervisord or systemd service templates to manage them.

**Comment by shshekhar93** [comment]
Fair enough, that makes sense. I agree this approach wouldn’t be appropriate for thousands of ports.

**Comment by gunzino** [comment]
@kzemek 
"Listening on thousand of ports" should be quite easy to achieve with small changes to the project. TPROXY iptables module can be used to direct all packets to the specific socket on the system with **transparent destination port and address**. The destination port can be then easily get from the connection information itself and used to establish connection to the upstream. The possible destination ports can be limited in the command line parameters.
